xdsl/affine_ir.py

```diff
@@ -0,0 +1,259 @@
+
+
+class _AffineExprKind(Enum):
+    """Enum for the kind of storage node used in AffineExpr."""
+
+    Add = auto()
+    Mul = auto()
+    Mod = auto()
+    FloorDiv = auto()
+    CeilDiv = auto()
+    Constant = auto()
+    DimId = auto()
+    SymbolId = auto()
+
+    def get_token(self):
+        """Get the token corresponding to the node kind."""
+        match self:
+            case _AffineExprKind.Add:
+                return "+"
+            case _AffineExprKind.Mul:
+                return "*"
+            case _AffineExprKind.Mod:
+                return "mod"
+            case _AffineExprKind.FloorDiv:
+                return "floordiv"
+            case _AffineExprKind.CeilDiv:
+                return "ceildiv"
+            case _AffineExprKind.Constant:
+                return "const"
+            case _AffineExprKind.DimId:
+                return "d"
+            case _AffineExprKind.SymbolId:
+                return "s"
+
+
+@dataclass
+class _AffineExprStorage:
+    """Base class for affine expression storage nodes."""
+
+    kind: _AffineExprKind
+
+
+@dataclass
+class _AffineBinaryOpExprStorage(_AffineExprStorage):
+    """An affine expression storage node representing a binary operation."""
+
+    lhs: AffineExpr
+    rhs: AffineExpr
+
+    def __post_init__(self) -> None:
+        if self.kind not in {
+            _AffineExprKind.Add,
+            _AffineExprKind.Mul,
+            _AffineExprKind.Mod,
+            _AffineExprKind.FloorDiv,
+            _AffineExprKind.CeilDiv,
+        }:
+            raise ValueError(f"Invalid kind {self.kind} for _AffineBinaryOpExprStorage")
+
+    def __str__(self) -> str:
+        return f"({self.lhs} {self.kind.get_token()} {self.rhs})"
+
+
+@dataclass
+class _AffineDimExprStorage(_AffineExprStorage):
+    """An affine expression storage node representing a dimension or symbol."""
+
+    position: int
+    """
+    The position of the dimension or symbol. Position of dimension and symbol
+    starts from 0 and is independent of each other. For example, if there are 2
+    dimensions and 3 symbols, then the positions of the dimensions are 0 and 1,
+    and the positions of the symbols are 0, 1, and 2.
+    """
+
+    def __post_init__(self) -> None:
+        if self.kind != _AffineExprKind.DimId and self.kind != _AffineExprKind.SymbolId:
+            raise ValueError(f"Invalid kind {self.kind} for _AffineDimExprStorage")
+        self.kind = self.kind
+        self.position = self.position
+
+    def __str__(self) -> str:
+        return f"{self.kind.get_token()}{self.position}"
+
+
+@dataclass
+class _AffineConstantExprStorage(_AffineExprStorage):
+    """An affine expression storage node representing a constant."""
+
+    value: int
+
+    def __init__(self, value: int) -> None:
+        self.kind = _AffineExprKind.Constant
+        self.value = value
+
+    def __str__(self) -> str:
+        return f"{self.value}"
+
+
+@dataclass()
+class AffineExpr:
```

Comment:
Can you add a bit of documentation on this class? Since this is the class that people will use.

Reply:
AffineExpr is not really meant to be used by anyone. I don't exactly know what to write. AffineExpr is not even Affine lol. I'll try to write something i guess.